### ISSUE TITLE
Associations linked by several "has_many, :through" show up blank in JSON.

### ISSUE THREAD

**Issue by nadeemja** [rails-api/active_model_serializers#1203]
Hi,

> TL;DR: Associations work perfectly when called from rails console using "Model.association", but records do not show up in JSON when defined in model_serializer as "has_many :association". What's up with that? :)

Here's the entire context:

I'm using active_model_serializers with rails_api. 

My model Event relates to the User model using a HABTM association.

For each Event, a User can have one of two roles: Lecturer or Student.

I manage these roles through the gem Rolify. 
https://github.com/RolifyCommunity/rolify

To get AMS and Rolify working together, I do this:

    # event.rb  
    class Event < ActiveRecord::Base
      (..)
      def lecturers
        User.with_role :lecturer, self
      end
    end

And,

    # event_serializer.rb
    class EventSerializer < ActiveModel::Serializer
      (...)
      has_many :lecturers
    end

This _works_. __But__, the :lecturers method is called for _each_ Event during the iteration. Which slows things down considerably. Since Rolify doesn't define Rails-relations, I can't do something like Event.includes(:lecturers).

__The solution__ that works in rails console, but _not_ in AMS, is to create models out of the tables Rolify uses, and link them using _has_many, :through_. Here's how I've done it:

    # models/role.rb
    # == Schema Information
    #
    # Table name: roles
    #
    #  id            :integer          not null, primary key
    #  name          :string(255)
    #  resource_id   :integer
    #  resource_type :string(255)
    #  created_at    :datetime
    #  updated_at    :datetime
    #

    class Role < ActiveRecord::Base
      has_and_belongs_to_many :users, :join_table => :users_roles
      belongs_to :resource, :polymorphic => true
      has_many :users_roles
      scopify
    end

And, 

    #models/users_role.rb 
    # == Schema Information
    #
    # Table name: users_roles
    #
    #  user_id :integer
    #  role_id :integer
    #

    class UsersRole < ActiveRecord::Base
      belongs_to :user
      belongs_to :role
    end

Finally, 

    #models/event.rb
    # == Schema Information
    #
    # Table name: events
    #
    #  id          :integer          not null, primary key
    #  title       :text
    #  description :text
    #  ext_id      :string(255)
    #  course_id   :integer
    #  created_at  :datetime
    #  updated_at  :datetime
    #  footnote    :string(255)
    #

    class Event < ActiveRecord::Base
      has_and_belongs_to_many :groups
      belongs_to :course
      belongs_to :type
      has_many :schedules
      has_many :attachments
      has_and_belongs_to_many :users 
      has_and_belongs_to_many :programs
      resourcify
      
      has_many :lecturer_roles, -> { where(name: 'lecturer', resource_type: 'Event' )}, class_name: 'Role', foreign_key: 'resource_id' 

      has_many :lecturer_users_roles, through: :lecturer_roles, source: :users_roles

      has_many :lecturer_users, class_name: 'User', through: :lecturer_users_roles, source: :user
      
      def lecturers

        User.with_role :lecturer, self
        
      end

      def students

        User.with_role :student, self
        
      end

    end

Yes, it's not pretty. 

However, in the rails console I can do: 

    Event.first.lecturer_users

Which not only works, but I can also do this now: 

    Event.includes(:lecturer_users)

Yet, in my event_serializer.rb, doing this:

    (...)
    has_many :lecturer_users
    (...)

Results in this in JSON:

    (...)
    lecturer_users: [ ]
    (...)

So, my question is: __Why does my implementation work in rails console, but not in AMS/rails_api?__

Thanks in advance, and thanks for the great work with AMS! :)

**Comment by beauby** [comment]
Which version of AMS are you using?

**Comment by nadeemja** [comment]
0.8.3. Have tried 0.10.x too.

**Comment by beauby** [comment]
Could you provide a github repo with a minimal project exhibiting the behavior?

**Comment by nadeemja** [comment]
Hmm.. Will try, but it might take a while. :)

**Comment by beauby** [comment]
One possible issue: you define a `lecturers` method on your model and a `has_many :lecturer_users` on your serializer. Have you made sure your naming is consistent?

**Comment by nadeemja** [comment]
Good question. The naming is consistent as far I can tell. I think I have tried with the lecturer method commented out, too. 

Thanks for looking into this. I will try to provide you with a minimal project as you requested.

**Comment by bf4** [comment]
Would you mind summarizing 
- the query (on Event) to eager load lecture users (or whatever the association is)
- the current state of the serializer

if what you need is `event = Event.includes(:lecturer_users).where(id: self.id); event.lecture_users`

**Comment by doei** [comment]
I've got a similar problem, namely I can't figure out how to customize nested has_many relations in my user_serializer.rb, and I also think it has something to do with my convoluted models.

So i've got something like this
My User can give a lecture, or receive a lecture
```ruby
#models/user.rb
#...
has_many :given_lecture, :class_name => 'Lecture', :foreign_key => 'teacher_id'
has_many :received_lecture, :class_name => 'Lecture', :foreign_key => 'student_id'

has_many :students, :through => :given_lectures
has_many :teachers, :through => :received_lectures
#...
```
All this works fine in console, and the serializer gets the corrects teachers and students, but i cannot edit the fields displayed.

Using github's AMS master branch I can edit nested fields of a simple model with a has_many relationship to another model:
```ruby
class ArticleSerializer < ActiveModel::Serializer
    attributes :content
    has_many :comments
    class CommentSerializer
        attributes :content
    end
end
```
That never worked for my User model, whatever AMS version I try. 

I also tried this syntax (found in some other issue):
```ruby
:has_many :comments do
    attributes :content
end
```
But it never worked either, whatever version I try


The workaround that currently saves me, is this one (also found in an issue), it doesn't work in the master branch version, only in 0.10.0.rc3:
Basically i'm forcing it to use my class

```ruby
#user_serializer.rb
class V1::UserSerializer < ActiveModel::Serializer
      has_many :students
#...
      def students
 	object.students.map { |x|
 		ActiveModel::Serializer::Adapter::FlattenJson.new(
 			StudentSerializer.new(x)).as_json}
       end

	class StudentSerializer < ActiveModel::Serializer
		attributes :id
	end
#...
end
```

Could you help me figure out what's going on?

**Comment by bf4** [comment]
@doei would you mind creating a new issue? Also, if you could include the link to issue that helped you and take a look at https://github.com/rails-api/active_model_serializers/blob/master/CONTRIBUTING.md ? Thanks!

**Comment by beauby** [comment]
Closing this for now. Feel free to reopen/keep commenting if needed.
